task_sdk/src/airflow/sdk/execution_time/supervisor.py

```diff
@@ -451,10 +449,38 @@ def _check_subprocess_exit(self):
 
     def _send_heartbeat_if_needed(self):
         """Send a heartbeat to the client if heartbeat interval has passed."""
-        if time.monotonic() - self._last_heartbeat >= FASTEST_HEARTBEAT_INTERVAL:
+        # If the process has exited, we don't need to send any more heartbeats
+        if self._exit_code is not None:
+            return
+
+        if time.monotonic() - self._last_heartbeat >= MIN_HEARTBEAT_INTERVAL:
             try:
                 self.client.task_instances.heartbeat(self.ti_id, pid=self._process.pid)
                 self._last_heartbeat = time.monotonic()
+            except ServerResponseError as e:
+                # TODO: Should we instead check httpx.HTTPStatusError?
+                if e.response.status_code == 409:
+                    reason = e.detail.get("reason", "")
+                    if reason == "not_running":
+                        error_msg = e.detail.get(
+                            "message", "Task is no longer in the running state and task should terminate"
+                        )
+                        log.error(error_msg, current_state=e.detail.get("current_state"))
+                    elif reason == "running_elsewhere":
+                        error_msg = e.detail.get("message", "Task is already running elsewhere")
+                        log.error(
+                            error_msg,
+                            current_hostname=e.detail.get("current_hostname"),
+                            current_pid=e.detail.get("current_pid"),
+                        )
+                elif e.response.status_code == 404:
+                    log.error("Task Instance not found")
+                else:
+                    # TODO: Handle other errors
+                    raise
+
+                # If heartbeating raises an error, kill the subprocess
+                self.kill(signal.SIGTERM)
```

Comment:
We don't want to kill the task on the first failure, we want to deal with temporary network glitches and blips and retry (next time around the loop, not a tenancity retry here I don't think) before killing the task proc

Reply:
added retry logic in https://github.com/apache/airflow/pull/44406/commits/d71cf35287fe01c3aa50bf2e0763c25ec44be061